js_modules/dagster-ui/packages/ui-core/src/runs/fetchPaginatedBucketData.ts

```diff
@@ -0,0 +1,67 @@
+export type QueryResultData<DataType> = {
+  data: DataType[] | undefined;
+  loading: boolean;
+  error: any;
+  called: boolean;
+};
+
+export async function fetchPaginatedBucketData<BucketType, DataType, CursorType, ErrorType>({
+  buckets,
+  fetchData,
+  setQueryData,
+}: {
+  buckets: BucketType[];
+  fetchData: (
+    bucket: BucketType,
+    cursor: CursorType | undefined,
+  ) => Promise<{
+    data: DataType[];
+    hasMore: boolean;
+    cursor: CursorType | undefined;
+    error: ErrorType;
+  }>;
```

Comment:
the fetchData function can be backed by a cache to avoid fetching previously fetched data. Not used in this PR since I need to figure out how to invalidate the cache when runs transition to completed and how to detect runs that transition to completed which I'm not sure is totally possible, or if it is it might only be possible to cache terminated run buckets